k4FWCore/components/Writer.cpp

```diff
@@ -0,0 +1,269 @@
+/*
+ * Copyright (c) 2014-2024 Key4hep-Project.
+ *
+ * This file is part of Key4hep.
+ * See https://key4hep.github.io/key4hep-doc/ for further info.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Gaudi/Functional/Consumer.h"
+#include "GaudiKernel/AnyDataWrapper.h"
+#include "GaudiKernel/IDataManagerSvc.h"
+#include "GaudiKernel/IDataProviderSvc.h"
+#include "GaudiKernel/SmartDataPtr.h"
+#include "GaudiKernel/StatusCode.h"
+
+#include "podio/Frame.h"
+
+#include "IIOSvc.h"
+
+#include "k4FWCore/DataWrapper.h"
+#include "k4FWCore/FunctionalUtils.h"
+
+#include <GaudiKernel/IHiveWhiteBoard.h>
+#include <memory>
+
+class Writer final : public Gaudi::Functional::Consumer<void(const EventContext&)> {
+public:
+  Writer(const std::string& name, ISvcLocator* svcLoc) : Consumer(name, svcLoc) {
+    // Non-reeentrant algorithms have a cardinality of 1
+    setProperty("Cardinality", 1).ignore();
+  }
+
+  mutable Gaudi::Property<std::vector<std::string>> m_OutputNames{this, "CollectionNames", {}};
+  mutable std::set<std::string>                     m_availableCollections;
+  mutable std::vector<std::string>                  m_collectionsToAdd;
+  mutable std::vector<std::string>                  m_collectionsRemaining;
+  mutable std::vector<std::string>                  m_collectionsToSave;
+
+  mutable std::mutex m_mutex;
+
+  ServiceHandle<IIOSvc>     iosvc{this, "IOSvc", "IOSvc"};
+  SmartIF<IHiveWhiteBoard>  m_hiveWhiteBoard;
+  SmartIF<IDataProviderSvc> m_dataSvc;
+  mutable bool              m_first{true};
+
+  StatusCode initialize() override {
+    if (!iosvc.isValid()) {
+      error() << "Unable to locate IIOSvc interface" << endmsg;
+      return StatusCode::FAILURE;
+    }
+
+    m_dataSvc = service("EventDataSvc", true);
+    if (!m_dataSvc) {
+      error() << "Unable to locate IDataSvc interface" << endmsg;
+      return StatusCode::FAILURE;
+    }
+
+    m_hiveWhiteBoard = service<IHiveWhiteBoard>("EventDataSvc", true);
```

Comment:
Different from `m_dataSvc` because it's a different type. From what I saw in the examples in Gaudi, when using the HiveWhiteBoard `EventDataSvc` is also used as the name of the service, so that means it can be any of the two.

Reply:
Let's just add some comments to clarify?